irods/files/rule-bases/ipc-encryption.re

```diff
@@ -0,0 +1,145 @@
+# Encryption enforcement policy
+
+
+# Checks if encryption is required for the collection entity
+_ipcIsEncryptionRequired(*Coll) {
+    *isRequired = false;
+    *res = SELECT META_COLL_ATTR_VALUE WHERE COLL_NAME == *Coll AND META_COLL_ATTR_NAME == 'encryption.required';
+    foreach (*record in *res) {
+        *isRequired = bool(*record.META_COLL_ATTR_VALUE);
+        break;
+    }
+    *isRequired;
+}
+
+
+# This rule checks if encryption is required and reject creating non-encrypted files
+_ipcEncryptionCheckEncryptionRequiredForDataObj(*Path) {
+    msiSplitPath(*Path, *parentColl, *objName);
+    # check if parent coll has encryption.required avu
+    if (_ipcIsEncryptionRequired(*parentColl)) {
+        # all encrypted files will have ".enc" extension
+        if (!cyverse_endsWith(*objName, ".enc")) {
+            # fail to prevent iRODS from creating the file without encryption
+            writeLine('serverLog', "Failed to create data object, encryption is required under *parentColl");
+            cut;
+            failmsg(-815000, 'CYVERSE ERROR:  attempt to create unencrypted data object');
+        }
+    }
+}
+
+_ipcEncryptionCheckEncryptionRequiredForCollInternal(*Coll) {
+    # check if src coll has non-encrypted data objects
+    *res = SELECT COLL_NAME, DATA_NAME WHERE COLL_NAME == *Coll;
+    foreach (*record in *res) {
+        # all encrypted files will have ".enc" extension
+        if (!cyverse_endsWith(*record.DATA_NAME, ".enc")) {
+            # fail to prevent iRODS from creating the file without encryption
+            writeLine('serverLog', "Failed to create data object, encryption is required under *Coll");
+            cut;
+            failmsg(-815000, 'CYVERSE ERROR:  attempt to create unencrypted data object');
+        }
+    }
+
+    *res = SELECT COLL_NAME WHERE COLL_PARENT_NAME == *Coll;
```

Comment:
To avoid recursion, a the set of child and grandchild collections can be gathered at once using a wildcard match. E.g. `SELECT COLL_NAME WHERE COLL_PARENT_NAME == *Coll || LIKE '*Coll/%'`